doc/design/tensor_design.md

```diff
@@ -0,0 +1,173 @@
+# Tensor Design
+
+
+## Allocation
+
+`Allocation` is a [RAII](http://en.cppreference.com/w/cpp/language/raii) class template, which is used to handle a piece of memory. 
+
+```cpp
+template <typename Device>
+class Allocation {
+ public:
+  Allocation();
+  Allocation(size_t size, Device device);
+
+  // Creates a non-owned allocation
+  Allocation(void *ptr, size_t size, Device device);
```

Comment:
What is the use of this interface? Tensor can share the pointer of the same Allocation object.